### PR TITLE
One more byte for inlined

### LINKED ISSUE #45: Maybe the `len` and enum branch can be merged into a `u8`?
Use 5 bit to save the length(max=31, enough for buffer), and use 2 bit to save the enum branch information(4 br). Thus the buffer size can be 23.
To implement it maybe we can no longer use `enum` and do transmute, the code may be a little bit ugly.

### PR DESCRIPTION
**Pull request by Austaras**
Closes #45

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -327,17 +328,52 @@ impl<'a> arbitrary::Arbitrary<'a> for SmolStr {
...
-const INLINE_CAP: usize = 22;
+const INLINE_CAP: usize = 23;
 const N_NEWLINES: usize = 32;
 const N_SPACES: usize = 128;
 const WS: &str =
     "\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n                                                                                                                                ";
 
+#[derive(Clone, Copy, Debug)]
+#[repr(u8)]
+enum InlineSize {
```
Comment (Austaras) [inline src/lib.rs:339]: Using macro or proc macro is possible -- however 24 isn't too many

File: src/lib.rs [hunk 2]
```diff
@@ -19,8 +20,8 @@ use core::{
 /// * `size_of::<SmolStr>() == size_of::<String>()`
 /// * `Clone` is `O(1)`
 /// * Strings are stack-allocated if they are:
-///     * Up to 22 bytes long
-///     * Longer than 22 bytes, but substrings of `WS` (see below). Such strings consist
+///     * Up to 23 bytes long
+///     * Longer than 23bytes, but substrings of `WS` (see below). Such strings consist
```
Comment (lnicola) [inline src/lib.rs:24]:
```suggestion
///     * Longer than 23 bytes, but substrings of `WS` (see below). Such strings consist
```

File: src/lib.rs [hunk 3]
```diff
@@ -51,16 +52,16 @@ impl SmolStr {
                 buf[$idx] = byte
             });
         }
-        s!(0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21);
+        s!(0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22);
         SmolStr(Repr::Inline {
-            len: len as u8,
+            len: unsafe { transmute(len as u8) },
```
Comment (Austaras) [inline src/lib.rs:57]: If `unsafe` should be avoided, a big match is also possible, and they generate identical code if opt level >= 1

Comment (lnicola) [reply src/lib.rs:57]: It might be nicer, but I'm not sure it's worth the hassle for `dev` builds. r? @Veykril

**Comment by Austaras** [comment]
Done

**Comment by lnicola** [comment]
The README also needs to be updated.

**Comment by Austaras** [comment]
> Do we make use of a niche introduced by it (if yes where)?

Yes, exactly. It's very simple in this case as other two variants only take two usize, so the first u8 can be used as tag. So if the first byte is 0-24, it's inlined string, 25 is heap string and 26 is space.

File: src/lib.rs (hunk 3 above)
Comment (Veykril) [reply src/lib.rs:57]: I think I prefer the match (if we have the option to avoid unsafe we should probably go for it)

File: src/lib.rs (hunk 1 above)
Comment (Veykril) [reply src/lib.rs:339]: Variants are guaranteed to pick the previous ones value + 1, so just specifying the first (for good measure) is enough

File: src/lib.rs (hunk 3 above)
Comment (Austaras) [reply src/lib.rs:57]: Well I checked again with something like
```rust
fn from_u8(input: u8) -> Self {
        match input {
            u if u == Self::_V0 as u8 => Self::_V0,
            // ...
            _ => unreachable!(),
        }
    }
```
and it generates
```asm
cmp     r12b, 23
mov     eax, 23
cmovb   eax, r12d
```
while `transmute` would simply be
```asm
mov     eax, r12d
```
So this unsafe is (at this point) unavoidable

**Comment by lnicola** [comment]
@Veykril I think we could merge this?

File: tests/test.rs [hunk 4]
```diff
@@ -43,12 +43,12 @@ fn old_const_fn_ctor() {
...
+    const LONG: SmolStr = SmolStr::new_inline_from_ascii(23, b"ABCDEFGHIZKLMNOPQRSTUVW");
 
     assert_eq!(EMPTY, SmolStr::from(""));
     assert_eq!(A, SmolStr::from("A"));
     assert_eq!(HELLO, SmolStr::from("HELLO"));
-    assert_eq!(LONG, SmolStr::from("ABCDEFGHIZKLMNOPQRSTUV"));
+    assert_eq!(LONG, SmolStr::from("ABCDEFGHIZKLMNOPQRSTUVW"));
 }
 
 fn check_props(std_str: &str, smol: SmolStr) -> Result<(), proptest::test_runner::TestCaseError> {
```
Comment (Veykril) [inline tests/test.rs:54]: This has a heap alloc check on lines 58/59, that should have the len check adjusted as well

**Comment by Veykril** [comment]
That one test should be adjusted, then I think it's good to merge

**Review by Veykril — APPROVED** [review]

**Pull request merged**